Documentation/postmortems/v3.5-data-inconsistency.md

```diff
@@ -0,0 +1,138 @@
+
+## Background
+
+Etcd v3 state is preserved on disk in two forms write ahead log (WAL) and database state (DB).
+Etcd v3.5 also still maintains v2 state, however it's deprecated and not relevant to the issue in this postmortem. 
+
+WAL stores history of changes for etcd state and database represents state at one point. 
+To know which point of history database is representing, it stores consistent index (CI).
+It's a special metadata field that points to last entry in WAL that it has seen.
+
+When Etcd is updating database state, it replays entries from WAL and updates the consistent index to point to new entry.
+This operation is required to be [atomic](https://en.wikipedia.org/wiki/Atomic_commit). 
+A partial fail would mean that database and WAL would no longer match, so some entries would be either skipped (if only CI is updated) or executed twice (if only changes are applied).
+This is especially important for distributed system like etcd, where there are multiple cluster members, each applying the WAL entries to their database.
+Correctness of the system depends on assumption that every member of the cluster, while replying WAL entries, will reach the same state.
+
+## Root cause
+
+To simplify managing consistency index, etcd has introduced backend hooks in https://github.com/etcd-io/etcd/pull/12855.
+Goal was to ensure that consistency index is always updated, by automatically triggering update during commit.
+Implementation was as follows, before applying the WAL entries, etcd updated in memory value of consistent index. 
+As part of transaction commit process, a database hook would read the value of consistent index and store it to database. 
+
+Problem is that in memory value of consistent index is shared, and there might be other in flight transactions apart from serial WAL apply flow.
+So if we imagine scenario:
+1. Etcd server starts an apply workflow, and it just sets a new consistent index value.
+2. The periodic commit is triggered, and it executes the backend hook and saves consistent index from apply workflow.
+3. Etcd server finished an apply workflow, saves new changes and saves same value of consistent index again.
+
+Between second and third point there is a very small window where consistent index is increased without applying any entries from WAL.
+
+## Trigger
+
+If etcd crashed after consistency index is saved, but before to apply workflow finished it would lead to data inconsistency.
+When recovering the data etcd would skip executing changes from failed apply workflow, assuming they have been already executed.
+
+This follows the issue reports and code used to reproduce the issue where trigger was etcd crashing under high request load.
+Crashes in reports where caused by etcd running under high memory pressure, causing it to go out of memory from time to time.
```

Comment:
The other source of crashing in 3.5.0 was: https://github.com/etcd-io/etcd/pull/13541